CASSANDRA-20982: Restrict BytesType value compatibility to scalar types, exclude collections and UDTs

Dekrate opened the pull request:
Restrict BytesType compatibility to scalar types only

BytesType.isValueCompatibleWith() previously accepted all types,
including frozen collections and UDTs. This was incorrect because
converting collections or UDTs to raw bytes is nonsensical - it
allows reading raw serialized bytes which have no meaningful
interpretation.

This patch restricts BytesType to only accept simple scalar types
by checking !isCollection() && !isUDT() in isValueCompatibleWithInternal().

@driftx <- it was your task, I was happy to help!

Dekrate committed:
Restrict BytesType compatibility to scalar types only

BytesType.isValueCompatibleWith() previously accepted all types,
including frozen collections and UDTs. This was incorrect because
converting collections or UDTs to raw bytes is nonsensical - it
allows reading raw serialized bytes which have no meaningful
interpretation.

This patch restricts BytesType to only accept simple scalar types
by checking !isCollection() && !isUDT() in isValueCompatibleWithInternal().

patch by Dekrate for CASSANDRA-20982

smiklosovic commented on test/unit/org/apache/cassandra/cql3/validation/operations/AlterTest.java:
```diff
@@ -68,15 +68,16 @@ public void testNonFrozenCollectionsAreIncompatibleWithBlob() throws Throwable
             createTable("CREATE TABLE %s (a int, b " + type + ", PRIMARY KEY (a));");
             alterTable("ALTER TABLE %s DROP b;");
-            alterTable("ALTER TABLE %s ADD b blob;");
+            assertInvalidMessage("Cannot re-add previously dropped column 'b' of type blob, incompatible with previous type " + type,
```
@Dekrate 

this reads like dropped column 'b' was of type blob while it was actually frozen. It would be more natural to say

`Cannot add a column 'b' of type blob, incompatible with previously dropped column 'b' of type frozen ...`

Dekrate replied:
Will fix it!

smiklosovic commented on src/java/org/apache/cassandra/db/marshal/BytesType.java:
```diff
@@ -90,8 +90,9 @@ public boolean isCompatibleWith(AbstractType<?> previous)
-        return true;
+        // BytesType should only be compatible with simple scalar types, not with collections or UDTs
+        // because converting a collection or UDT to raw bytes is nonsensical
+        return !otherType.isCollection() && !otherType.isUDT();
```
`!otherType.isMultiCell()`?

Dekrate committed:
Use isMultiCell() and improve error message clarity

Replace isCollection() && isUDT() checks with isMultiCell() for cleaner
code. Update error message in AlterTableStatement to more clearly indicate
that we're adding a column of incompatible type with a previously dropped
column, rather than suggesting the dropped column had the new type.

Update tests to match the new error message format.

patch by Dekrate for CASSANDRA-20982

Dekrate commented:
@smiklosovic, done

driftx commented on src/java/org/apache/cassandra/db/marshal/BytesType.java:
```diff
@@ -90,8 +90,9 @@ public boolean isCompatibleWith(AbstractType<?> previous)
-        return true;
+        // BytesType should only be compatible with simple scalar types, not with collections or UDTs
+        // because converting a collection or UDT to raw bytes is nonsensical
+        return !otherType.isMultiCell();
```
Doing this here will break cql CAST operations, which I think we still might want to allow since they are explicit, and different from finding a column you re-added contains junk bytes.

Also, `isMultiCell` will return false for frozen collections.

Dekrate replied:
@driftx, should I revert to the first version?

`return !otherType.isCollection() && !otherType.isUDT();`

driftx replied:
Yes and we shouldn't do this in `isValueCompatibleWithInternal` since it will prevent CASTing to bytes.

Dekrate replied:
@driftx, thanks! Could you please suggest any particular place to put it in?

To sum up: 
1. `isValueCompatibleWithInternal` returns true back
2. `return !otherType.isCollection() && !otherType.isUDT();` will be in the other place?

Dekrate replied:
Maybe this `isSerializationCompatibleWith`?

driftx replied:
`isSerializationCompatibleWith` will prevent the schema change but still allow the cast.

Dekrate committed:
Restrict BytesType ALTER TABLE compatibility to scalar types

BytesType.isSerializationCompatibleWith() now rejects collections
and UDTs to prevent nonsensical schema changes, while still allowing
CAST operations through isValueCompatibleWith().

patch by Dekrate for CASSANDRA-20982

Dekrate commented:
@driftx, resolved.

Dekrate commented:
Any updates @driftx @smiklosovic? I've seen the discussion on Jira platform.

Dekrate committed:
Changed testInvalidDroppingAndAddingOfCollections to support new exception message.

patch by Dekrate for CASSANDRA-20982
